zkevm-circuits/src/evm_circuit/op_execution/utils/math_gadgets.rs

```diff
@@ -83,3 +85,146 @@ impl<F: FieldExt> IsEqualGadget<F> {
         self.is_zero.assign(region, offset, lhs - rhs)
     }
 }
+
+/// Returns `1` when `lhs < rhs`, and returns `0` otherwise.
+/// lhs and rhs `< 256**NUM_BYTES`
+/// `NUM_BYTES` is required to be `<= MAX_BYTES_FIELD` to prevent overflow:
+/// values are stored in a single field element and two of these
+/// are added together.
+/// The equation that is enforced is `lhs == rhs + diff - (lt * range)`.
+/// Because all values are `<= 256**NUM_BYTES` and `lt` is boolean,
+///  `lt` can only be `1` when `lhs < rhs`.
+#[derive(Clone, Debug)]
+pub struct LtGadget<F, const NUM_BYTES: usize> {
+    lt: Cell<F>, // `1` when `lhs < rhs`, `0` otherwise.
+    diff: [Cell<F>; NUM_BYTES], /* The byte values of `diff`.
+                  * `diff` equals `lhs - rhs` if `lhs >= rhs`,
+                  * `lhs - rhs + range` otherwise. */
+    range: F, // The range of the inputs, `256**NUM_BYTES`
+}
+
+impl<F: FieldExt, const NUM_BYTES: usize> LtGadget<F, NUM_BYTES> {
+    pub const NUM_CELLS: usize = 1 + NUM_BYTES; // lt + diff
+    pub const NUM_WORDS: usize = 0;
+
+    pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
+        assert!(NUM_BYTES <= MAX_BYTES_FIELD, "unsupported number of bytes");
+        Self {
+            lt: alloc.cells.pop().unwrap(),
+            diff: array_init(|_| alloc.cells.pop().unwrap()),
+            range: utils::get_range(NUM_BYTES * 8),
+        }
+    }
+
+    pub(crate) fn constraints(
+        &self,
+        cb: &mut ConstraintBuilder<F>,
+        lhs: Expression<F>,
+        rhs: Expression<F>,
+    ) -> Expression<F> {
+        let diff = utils::from_bytes::expr(self.diff.to_vec());
+        // The equation we require to hold: `lhs == rhs + diff - (lt * range)`.
+        cb.require_equal(lhs, rhs + diff - (self.lt.expr() * self.range));
+
+        // `lt` needs to be boolean
+        cb.require_boolean(self.lt.expr());
+        // All parts of `diff` need to be bytes
+        for byte in self.diff.iter() {
+            cb.require_in_range(byte.expr(), 256);
+        }
+
+        self.lt.expr()
+    }
+
+    pub(crate) fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        lhs: F,
+        rhs: F,
+    ) -> Result<(F, Vec<u8>), Error> {
+        // Set `lt`
+        let lt = lhs < rhs;
+        self.lt.assign(
+            region,
+            offset,
+            Some(F::from_u64(if lt { 1 } else { 0 })),
+        )?;
+
+        // Set the bytes of diff
+        let diff = (lhs - rhs) + (if lt { self.range } else { F::zero() });
+        let diff_bytes = diff.to_bytes();
```

Comment:
One question here is that I remember previously @han0110 and @miha-stopar thinks the intermediate results should still be provided in the witness instead of computing it in the assignment. Should we still follow this principle?

Reply:
Hmm I don't believe that would make things better. All these intermediate results are implementation dependent, while the actual input witness for an opcode would best be as general as possible to allow the circuits to use that data as needed. Let's say we change the implementation to work on 2 bytes instead of 1 after we add a `range65536` lookup table to optimize things, I would expect this to not impact the witness format generated by the busmapping.

Reply:
The reason we don't feed the `geth` trace into circuit directly is because we need to pre-process to get some metadata (like counter, result in the future, etc...), then we thought maybe we could just generate everything in pre-processing stage including "intermediate witness".

However, after more opcodes are implemented, such "intermediate witness" seems hard to generated from crate `bus-mappping` since their type and amount could be very different, and it becomes even trivial for those opcodes with complicated logics. So I feel that @Brechtpd's idea makes more sense to me now.

Maybe we should just let `bus-mappping` provide the `Operation` which contains all necessary kinds of read/write (which is just current API), and let circuit generate "intermediate witness" itself for flexibility. What do you think @CPerezz, @ed255?